cmake: Split toolchain configuration into pre-and post-DT

SebastianBoe opened the pull request:
Split up the toolchain configuration into two phases, generic and
target. The 'generic' phase configures the toolchain just enough to be
able to preprocess DT files. The 'target' phase completes the
configuration with target-specific configuration.

SebastianBoe commented:
I would like to see this regression tested against all toolchains, but don't want to install all toolchains locally. Is this possible?

carlescufi commented:
@SebastianBoe isn't the Zephyr SDK enough? That contains toolchains for all architectures

galak reviewed (changes requested):
I built with arm embedded, xtensa-gcc, and tried with xtools.

The xtools fails for x86 (added a comment about prefix changing).

SebastianBoe commented on cmake/toolchain/xtools/generic.cmake:
```diff
@@ -0,0 +1,41 @@
+
+set(COMPILER gcc)
+
+if("${ARCH}" STREQUAL "x86")
```
It comes from here: https://github.com/zephyrproject-rtos/zephyr/pull/11810/files/9360f938aaec559d03787762237164f46899258e..9bb17be2b79ed6d8e5852105c9de57f36e5734fe#diff-4b7d80e654647222dd0d890357d37b9aL10

Now that we can't access CONFIG_ we check the existence of the toolchain instead.

Will look into the xtools x86 failure.

SebastianBoe replied:
I was not able to reproduce the x86 xtools failure.

This is my environment:

export ZEPHYR_SDK_INSTALL_DIR=~/zephyr-sdk
export XTOOLS_TOOLCHAIN_PATH="~/x-tools"
export ZEPHYR_TOOLCHAIN_VARIANT=xtools
export BOARD=qemu_x86
cmake -H~/zephyr/samples/hello_world -Boe

SebastianBoe replied:
Please post steps to reproduce.

SebastianBoe replied:
Fixed.

SebastianBoe commented:
> The xtools fails for x86 (added a comment about prefix changing).
-- @galak 

Fixed.

SebastianBoe commented:
> The xtools fails for x86 (added a comment about prefix changing).
-- @galak 

Fixed.

SebastianBoe commented:
> The xtools fails for x86 (added a comment about prefix changing).
-- @galak 

Fixed.

galak commented:
@mped-oticon can you take a look at this.

mped-oticon commented on cmake/compiler/gcc/target.cmake:
```diff
@@ -1,8 +1,11 @@
-set_ifndef(CC gcc)
 set_ifndef(C++ g++)
```
for consistency, I think it would be better to also removing if-not-def'd C++ too.

SebastianBoe replied:
This would violate the design.

generic.cmake contains just enough configuration to run a preprocessor on DT files. And target.cmake contains the rest.

mped-oticon commented on cmake/compiler/gcc/target.cmake:
```diff
@@ -1,8 +1,11 @@
 # GCC-based toolchains
+
+unset(CMAKE_C_COMPILER CACHE)
+unset(CMAKE_C_COMPILER)
```
These unsets would be needed in every `target.cmake`. Factor out so it becomes shared.

mped-oticon commented on cmake/toolchain/clang/generic.cmake:
```diff
@@ -0,0 +1,5 @@
+set_ifndef(CLANG_ROOT /usr)
+set(COMPILER clang)
+
+set(CMAKE_C_COMPILER          ${CLANG_ROOT}/bin/clang)
```
setting `CMAKE_C_COMPILER` does not belong to a toolchain file, it belongs to the `cmake/compiler/${COMPILER}/{target,generic}.cmake` files.
To be clear, the toolchain files should set the `COMPILER` variable pointing to the cmake file responsible for finding and setting the actual compiler (`CMAKE_C_COMPILER`).
That two-stage rocket is a nice approach, and we will eventually follow the same for linker and binutils.

mped-oticon replied:
clang is not a toolchain. Strictly speaking, clang is a parser.

It would clear up confusion if we named according to:

- `llvm` as the toolchain (clang, lld, libs, tools).
- `clang` as the compiler.

SebastianBoe replied:
I agree with all points, but due to problems with the clang toolchain[0]. I would prefer not to resolve these issues in this PR.

Note that these comments apply equally well before and after this PR, so they are out-of-scope of this PR.

[0] https://github.com/zephyrproject-rtos/zephyr/issues/11187#issuecomment-445245153

mped-oticon commented on cmake/toolchain/xtools/generic.cmake:
```diff
@@ -0,0 +1,23 @@
+set_ifndef(XTOOLS_TOOLCHAIN_PATH "$ENV{XTOOLS_TOOLCHAIN_PATH}")
+set(       XTOOLS_TOOLCHAIN_PATH    ${XTOOLS_TOOLCHAIN_PATH} CACHE PATH "")
+assert(    XTOOLS_TOOLCHAIN_PATH     "XTOOLS_TOOLCHAIN_PATH is not set")
```
spacing

SebastianBoe replied:
Fixed (vertically aligned the code for readability)

mped-oticon commented on cmake/toolchain/zephyr/generic.cmake:
```diff
@@ -0,0 +1,8 @@
+# Ensures toolchain-gcc.cmake is run
```
stale comment

SebastianBoe replied:
Fixed here: https://github.com/zephyrproject-rtos/zephyr/pull/11942/files#diff-801abdd2ea4d62949a262ba3cae57e2cL3

mped-oticon commented on cmake/target_toolchain.cmake:
```diff
@@ -0,0 +1,31 @@
+
+# Configure the toolchain based on what toolchain technology is used
+# (gcc, host-gcc etc.)
+include(${ZEPHYR_BASE}/cmake/compiler/${COMPILER}/target.cmake OPTIONAL)
```
Dissonance between base directories for compiler and toolchain folders:
Above for host-gcc, `TOOLCHAIN_ROOT` was used, but here we use `ZEPHYR_BASE`.
Should be `TOOLCHAIN_ROOT` in both places as that works with out-of-tree stuff.

mped-oticon replied:
(`TOOLCHAIN_ROOT` is btw ill-named as it is not the root dir of any toolchain -- just the dir holding some cmake files for toolchains.)

SebastianBoe replied:
What if you want to combine a compiler from Zephyr with an out-of-tree toolchain?

I suspect this code is working-as-intended.

mped-oticon replied:
Even if just for consistency, they should be the same.
I believe Oticon is currently the main (only?) user of out-of-tree due to commercial custom toolchain. 

Anyway, I can fix this later.

mped-oticon commented on cmake/generic_toolchain.cmake:
```diff
@@ -56,30 +46,16 @@ assert(TOOLCHAIN_ROOT "Zephyr toolchain root path invalid: please set the TOOLCH
 assert(ZEPHYR_TOOLCHAIN_VARIANT "Zephyr toolchain variant invalid: please set the ZEPHYR_TOOLCHAIN_VARIANT-variable")
 
-if(CONFIG_ARCH_POSIX OR (ZEPHYR_TOOLCHAIN_VARIANT STREQUAL "host"))
+if(${ARCH} STREQUAL "posix" OR (ZEPHYR_TOOLCHAIN_VARIANT STREQUAL "host"))
```
Separate PR?
If you're changing this anyway, why not clean it up so this special case is eliminated altogether?

SebastianBoe replied:
Will close this, do a separate PR, and then re-open.

SebastianBoe replied:
> If you're changing this anyway, why not clean it up so this special case is eliminated altogether?

This is out-of-scope of this PR. But a PR that did this would be welcome.

mped-oticon commented on cmake/generic_toolchain.cmake:
```diff
@@ -56,30 +46,16 @@ assert(TOOLCHAIN_ROOT "Zephyr toolchain root path invalid: please set the TOOLCH
-
-file(MD5 ${CMAKE_C_COMPILER} CMAKE_C_COMPILER_MD5_SUM)
-set(TOOLCHAIN_SIGNATURE ${CMAKE_C_COMPILER_MD5_SUM})
+include(${ZEPHYR_BASE}/cmake/compiler/${COMPILER}/generic.cmake OPTIONAL)
```
`TOOLCHAIN_ROOT` like above.

SebastianBoe replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/11810#discussion_r239779050

mped-oticon commented on cmake/app/boilerplate.cmake:
```diff
@@ -256,8 +256,22 @@ set(CMAKE_CXX_COMPILER_FORCED 1)
+# DTS should be close to kconfig because CONFIG_ variables from
+# kconfig and dts should be available at the same time.
+#
+# The DT system uses a C preprocessor for it's code generation needs.
```
@galak Does DTS actually require the target-toolchain to run? Would plain gcc not be sufficient if it only requires the preprocessor?

If so, then we could avoid O(n)-many `generic.cmake` files.

SebastianBoe replied:
It doesn't require the target-toolchain. It requires 'any' C toolchain (preprocessor).

'gcc' is not sufficient. Because 'gcc' is not a dependency, so we can't avoid O(n)-many generic.cmake files.

Said differently, Zephyr does not require a host C toolchain to be installed, but can rely solely on a single cross-compiling toolchain.

SebastianBoe replied:
IMHO, since the C toolchain is just used as a code generator, we should long-term migrate DT to use Jinja instead of CPP, and merge back together the toolchain config files. But this will have to wait for now ...

SebastianBoe replied:
Yes.

Technically #include is not needed because DT has include, but it does have it.

http://jinja.pocoo.org/docs/2.10/templates/#include
http://jinja.pocoo.org/docs/2.10/templates/#assignments
http://jinja.pocoo.org/docs/2.10/templates/#macros

galak replied:
I see, will I don't think we will move to using jinja for DTS files as we want to share DTS with Linux Kernel or other systems.

mped-oticon replied:
@SebastianBoe True, but ZephyrSDK is a dependency and there we pack a couple of gcc's. We could pick a preprocessor from there?

SebastianBoe replied:
>  ZephyrSDK

Not for all platforms.

mped-oticon reviewed (changes requested):
Good to have some more rigour.

However, as I've commented, having double the toolchains (previously we had only `target`, now there is also `generic` due to DTS) may be overkill if we only need the preprocessor for DTS.

BTW, this change breaks out-of-tree toolchains due to new expectations (fine with me), but should be clearly signalled.

nashif commented:
> BTW, this change breaks out-of-tree toolchains due to new expectations (fine with me), but should be clearly signalled.

we probably need a test for out-of-tree toolchains, no test, no guarantee a certain feature will continue to work, this is the sad reality and we need to improve this.

SebastianBoe commented on cmake/compiler/gcc/generic.cmake:
```diff
@@ -0,0 +1,26 @@
+  message(FATAL_ERROR "Executing the below command failed. Are permissions set correctly?
+'${CMAKE_C_COMPILER} --version'
+"
+    )
```
Hi, please clarify what is wrong with the spacing.

My editor is inserting 2 spaces for indentation. And indents closing parenthesis in function calls. This gives 4 spaces in total in this case.

SebastianBoe commented:
Temporarily closing the PR: https://github.com/zephyrproject-rtos/zephyr/pull/11810#discussion_r239772539

SebastianBoe committed:
cmake: Split toolchain configuration into pre-and post-DT

Split up the toolchain configuration into two phases, generic and
target. The 'generic' phase configures the toolchain just enough to be
able to preprocess DT files. The 'target' phase completes the
configuration with target-specific configuration.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>

carlescufi commented:
@mped-oticon are you happy with this now as-is?
